Bluetooth: Controller: Broadcast ISO encryption support

cvinayak opened the pull request:
Add implementation to parse ACAD and find the BIGInfo before
setting up Broadcast ISO Synchronization.

To be consistent in naming as len instead of length in PDU
struct definitions.

Implementation of Broadcast ISO encryption using h8 function

Fixes #41955
Fixes #41956

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add packing sequential or interleaved comment

Add code comment related to BIG packing sequential or
interleaved being used.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Use len instead of length in PDU structs

To be consistent in naming as len instead of length in PDU
struct definitions.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Parse and find BIGInfo to setup ISO sync

Add implementation to parse ACAD and find the BIGInfo before
setting up Broadcast ISO Synchronization.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Broadcast ISO encryption support

Implementation of Broadcast ISO encryption using crypto
toolbox function h8. And support for encryption in lower
link layer for nRF5x series.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:
```diff
@@ -345,8 +343,33 @@ static int prepare_cb_common(struct lll_prepare_param *p)
 	pdu->cssn = lll->cssn;
 
-	radio_pkt_tx_set(pdu);
+	/* Encryption */
+	if (pdu->len && lll->enc) {
+		uint8_t pkt_flags;
 
+		lll->ccm_tx.counter = payload_count;
+
+		(void)memcpy(lll->ccm_tx.iv, lll->giv, 4U);
+		mem_xor_32(lll->ccm_tx.iv, lll->ccm_tx.iv, access_addr);
+
+		pkt_flags = RADIO_PKT_CONF_FLAGS(RADIO_PKT_CONF_PDU_TYPE_DC,
+						 phy,
+						 RADIO_PKT_CONF_CTE_DISABLED);
+		radio_pkt_configure(RADIO_PKT_CONF_LENGTH_8BIT,
+				    (lll->max_pdu + PDU_MIC_SIZE), pkt_flags);
+		radio_pkt_tx_set(radio_ccm_tx_pkt_set(&lll->ccm_tx, pdu));
+	} else {
+		uint8_t pkt_flags;
+
+		pkt_flags = RADIO_PKT_CONF_FLAGS(RADIO_PKT_CONF_PDU_TYPE_DC,
+						 phy,
+						 RADIO_PKT_CONF_CTE_DISABLED);
+		radio_pkt_configure(RADIO_PKT_CONF_LENGTH_8BIT, lll->max_pdu,
+				    pkt_flags);
+		radio_pkt_tx_set(pdu);
+	}
+
+	/* Setup radio IFS switching */
```
Define RADIO_PKT_CONF_PDU_TYPE_BIS and RADIO_PKT_CONF_PDU_TYPE_CIS, and use it to set correct PDU header masking required for CCM specification's additional authentication data.

cvinayak commented:
@niag-Oticon @wopu-ot Please provide your comments, do review the use of `len` in this commit too https://github.com/zephyrproject-rtos/zephyr/pull/50926/commits/76f8caa37232d538dff15e086ebe8accf039dc39

wopu-ot commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -873,7 +873,13 @@ struct pdu_iso {
+
+	uint8_t len;
+
+	union {
+		uint8_t resv;
+	};
```
This is a non-standard field that is specific to the Nordic radio and should not be unconditionally part of the definition of a PDU. (It is bad enough that it is part of some existing definitions, but we should at least not make the problem worse.)

cvinayak replied:
Would it be acceptable to have a `pdu_vendor.h` with:

```c
/* Presence of vendor PDU struct octet3 */
struct pdu_data_vnd_octet3 {
	union {
		uint8_t resv;
	};
} __packed;
```

or

```c
/* No vendor PDU struct octet3*/
struct pdu_data_vnd_octet3 {
	uint8_t resv[0];
} __packed;
```

and include:

```c
	struct pdu_data_vnd_octet3 vnd_octet3;
```

wopu-ot replied:
Having a `pdu_vendor.h` sounds fine to me. However, putting a zero-length array in the middle of another struct won't fly due to the C standard. Maybe a define like `PDU_DATA_VENDOR_EXT_HEADER` could be an option?

cvinayak replied:
GCC built fine when I have a zero-length array in between in the pdu_data struct, we have zero-length array at end of struct allover HCI too.

carlescufi replied:
> Having a `pdu_vendor.h` sounds fine to me. However, putting a zero-length array in the middle of another struct won't fly due to the C standard. Maybe a define like `PDU_DATA_VENDOR_EXT_HEADER` could be an option?

How so? Is this actually forbidden somewhere?

cvinayak replied:
@wopu-ot With the direction finding extensions, the 3rd Octet (optional) is a BT Standard for data channel PDUs. Should we just agree to keep the existing union? Is it an issue for your implementation (the octet is already upstream)?

wopu-ot replied:
@carlescufi I am afraid I confused the rules for flexible-length arrays and zero-length arrays. Zero-length arrays in the middle of things are not a problem from the standard's perspective as far as I understand now. Sorry for the confusion.

wopu-ot replied:
@cvinayak I would be happy to get rid of the additional byte, but we can live with what is upstream now. I would like to avoid making the issue worse by adding the `resv` field in more places though.

cvinayak replied:
@wopu-ot  Is pdu_vendor.h an acceptable solution, vendors can omit the octet3. See, https://github.com/zephyrproject-rtos/zephyr/pull/50926/commits/749e67597c2cd1230588af6f3b7fd6e808d3bacb#diff-8d21a201cb026a71cf58c6f34d9ee48816ff8ec493cb6426d02be884bc498b8eR1

wopu-ot replied:
I believe the pdu_vendor.h solution should work for our end.

niag-Oticon reviewed (commented):
LGTM but I'm not familiar with much of the affected areas apart from the ISO-AL.

cvinayak committed:
Bluetooth: Controller: Make PDU octet3 vendor specific

Make Data, ISO, BIS and CIS PDU structure's octet 3 vendor
specific.

This will allow vendors not supporting the octet 3 or
CTEInfo (8 bits) to save 1 octet in their PDU structures.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Fix endianness of cryptographic toolbox function h8

Convert endianness of supplied parameters before calculating
the AES-CMAC. Bluetooth stores values in little-endian and
crypto traditionally operates on big-endian storage.

Relates to commit e9c542ab5b48 ("Bluetooth: Add the
cryptographic toolbox function h8").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

wopu-ot reviewed (commented):
I believe the pdu_vendor.h solution should work for our end.

cvinayak commented:
Closing this PR, will be superceded by a consolidated PR with all ISO implementation. The PR will probably be https://github.com/zephyrproject-rtos/zephyr/pull/52842

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -397,6 +401,43 @@ uint8_t ll_big_create(uint8_t big_handle, uint8_t adv_handle, uint8_t num_bis,
+
+		/* Prepare the CCM parameters */
+		ccm_tx = &lll_adv_iso->ccm_tx;
+		ccm_tx->direction = 1U;
```
Is `direction` like RX/TX or? Should it have values that are `#define`d?

Thalley replied:
Alright, so `1U` is sent from central, and `0U` is sent from peripheral? 

IMO that sounds like predefined values that should be put into properly defined values to make it more clear to the reader

cvinayak replied:
Hope it is ok to make changes outside this PR, as the changes would impact more that the Broadcast ISO related changes.

![image](https://user-images.githubusercontent.com/6350656/214570315-501e3c5b-7206-4528-b607-5e0bfdf224b3.png)

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync_iso.c:
```diff
@@ -142,6 +144,28 @@ uint8_t ll_big_sync_create(uint8_t big_handle, uint16_t sync_handle,
+
+		/* Prepare the CCM parameters */
+		ccm_rx = &lll->ccm_rx;
+		ccm_rx->direction = 1U;
```
Ditto

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync_iso.c:
```diff
@@ -334,13 +358,30 @@ void ull_sync_iso_setup(struct ll_sync_iso_set *sync_iso,
+			break;
+		}
+
+		acad_len -= hdr_len + PDU_ADV_DATA_HEADER_LEN_SIZE;
```
Any risk of underrun here?

cvinayak replied:
Added check before decrement.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync_iso.c:
```diff
@@ -384,6 +425,30 @@ void ull_sync_iso_setup(struct ll_sync_iso_set *sync_iso,
+
+		/* Prepare the CCM parameters */
+		ccm_rx = &lll->ccm_rx;
+		ccm_rx->direction = 1U;
+		(void)memcpy(&ccm_rx->iv[4], &lll->giv[4], 4U);
```
Ditto

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync_iso.c:
```diff
@@ -142,6 +144,28 @@ uint8_t ll_big_sync_create(uint8_t big_handle, uint16_t sync_handle,
 	lll->term_reason = 0U;
 
+	if (encryption) {
+		const uint8_t BIG1[16] = "1GIB";
+		const uint8_t BIG2[4] = "2GIB";
```
```suggestion
		const char BIG1[16] = "1GIB";
		const char BIG2[4] = "2GIB";
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync_iso.c:
```diff
@@ -384,6 +425,30 @@ void ull_sync_iso_setup(struct ll_sync_iso_set *sync_iso,
 	lll->payload_count |= (uint64_t)bi->payload_count_framing[4] << 32;
 
+	if (lll->enc && (bi_size == PDU_BIG_INFO_ENCRYPTED_SIZE)) {
+		const uint8_t BIG3[4] = "3GIB";
```
```suggestion
		const char BIG3[4] = "3GIB";
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync_iso.c:
```diff
@@ -384,6 +425,30 @@ void ull_sync_iso_setup(struct ll_sync_iso_set *sync_iso,
+	if (lll->enc && (bi_size == PDU_BIG_INFO_ENCRYPTED_SIZE)) {
+		const uint8_t BIG3[4] = "3GIB";
+		struct ccm *ccm_rx;
+		uint8_t gsk[16];
```
Should the sizes of the `gsk`, `igltk`, etc. be `#define`d somewhere? Perhaps in `include/zephyr/crypto/crypto.h`?

Thalley replied:
I've been adding it when implementing the crypto functions for CSIP: 
```c
#define BT_CSIP_CRYPTO_KEY_SIZE   16
#define BT_CSIP_CRYPTO_SALT_SIZE  16
```

Several other modules have done the same, e.g.
```c
#define ECB_AES_KEY_SIZE   16
#define ECB_AES_BLOCK_SIZE 16
```
and 
```c
#define ATAES132A_AES_KEY_SIZE 16
```

So while `smp.c` doesn't do it, we do it several other places, and I suggest that we do define and use them here as well, as it makes it painfully clear what the individual sizes represent.

cvinayak replied:
Most public headers too do not define array size as defines, example `bluetooth.h` (LE SC OOB data), `hci.h`, `crypto.h` etc.. as long as `sizeof()` or `ARRAY_SIZE` can be used, it is redundant to have a define for array size.

Thalley commented on subsys/bluetooth/controller/util/mem.c:
```diff
@@ -134,6 +134,24 @@ uint8_t mem_nz(uint8_t *src, uint16_t len)
+/**
+ * @brief XOR bytes
+ */
+inline void mem_xor_n(uint8_t *dst, uint8_t *src1, uint8_t *src2, uint16_t len)
+{
+	while (len--) {
+		*dst++ = *src1++ ^ *src2++;
+	}
+}
+
+/**
+ * @brief XOR 32-bits
+ */
+void mem_xor_32(uint8_t *dst, uint8_t *src1, uint8_t *src2)
+{
+	mem_xor_n(dst, src1, src2, 4U);
+}
```
These seem useful outside the scope of the BT controller. Should we consider adding them to a public API somewhere?

Thalley commented on subsys/bluetooth/crypto/bt_crypto.c:
```diff
@@ -261,18 +261,34 @@ int bt_crypto_h7(const uint8_t salt[16], const uint8_t w[16], uint8_t res[16])
+	LOG_DBG("s %s", bt_hex(s, 16));
+	LOG_DBG("key_id %s", bt_hex(key_id, 4));
+
+	sys_memcpy_swap(ks, k, 16);
+	sys_memcpy_swap(ss, s, 16);
```
Why are we swapping these? 

Should they be swapped for both big and little endian?

Thalley replied:
Have these functions been tested with a big endian device? 

The documentation for the `bt_crypto` functions does not specify anything regarding the endianess of the inputs, so it sounds like its mostly based on assumptions? 

In any case, you are being consistent with the rest of `bt_crypto.c`, so no change is required in this PR.

cvinayak replied:
hmmm... I could not find any reference/documentation in the Zephyr tree as to the endian-ness of stored keys and that Bluetooth PDU contents stored as protocol endian-ness.

For now, I will believe its assumptions in BT stack and not documenting the endian-ness of key storage.

cvinayak commented:
Re-opening to address change requests

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:
```diff
@@ -345,8 +344,33 @@ static int prepare_cb_common(struct lll_prepare_param *p)
 
+		lll->ccm_tx.counter = payload_count;
+
+		(void)memcpy(lll->ccm_tx.iv, lll->giv, 4U);
```
Fair. Does it make sense to use `sizeof(lll->ccm_tx.iv)`?

cvinayak replied:
No, lll->ccm_tx.iv  is 8 bytes.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_iso.c:
```diff
@@ -569,22 +598,22 @@ static void isr_tx_common(void *param,
 	if (!pdu) {
+		uint8_t payload_index;
+
+		payload_index = (lll->bn_curr - 1U) +
```
Might just be missing the bigger picture here (hard to get with larger PRs), but there's plenty of places in the code where `lll->bn_curr` is assigned `0U` as well. But that can never happen when we enter this function?

cvinayak replied:
hmmm.... fixed the assignments, they where pseudo code before (0 assigned and post incremented) ... fixed now.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/pdu_vendor.h:
```diff
@@ -0,0 +1,55 @@
+
+/* Presence of vendor BIS PDU struct octet3 */
+struct pdu_bis_vnd_octet3 {
+	union {
+		uint8_t resv[OCTET3_LEN]; /* nRF specific octet3 required for NRF_CCM use */
+	} __packed;
+} __packed;
+
+/* Presence of vendor CIS PDU struct octet3 */
+struct pdu_cis_vnd_octet3 {
+	union {
+		uint8_t resv[OCTET3_LEN]; /* nRF specific octet3 required for NRF_CCM use */
+	} __packed;
+} __packed;
+
+/* Presence of ISOAL helper vendor ISO PDU struct octet3 */
+struct pdu_iso_vnd_octet3 {
+	union {
+		uint8_t resv[OCTET3_LEN]; /* nRF specific octet3 required for NRF_CCM use */
+	} __packed;
```
I don't have a reference from the core spec to what the "3rd octet" is, so I can't really tell if it makes sense, but I assume it does :)

w.r.t. it being a nameless union now, adding the `union` later would be trivial, but at the same time it doesn't do any harm as it is.

cvinayak replied:
![image](https://user-images.githubusercontent.com/6350656/214552515-9e15b4f8-8c9a-4333-928c-2ac01dfb56a1.png)

cvinayak commented:
@wopu-ot  I had to re-open this PR as comments had to be addressed and hence to get CI passing. The major change needing some porting on your platforms would be the introduction of `pdu_df.h` and `pdu_vendor.h`. As I had prior agreement on the `pdu_vendor.h` approach, hope I can get an approval on this PR so as to make it for the feature freeze.

mtpr-ot reviewed (approved):
LGTM

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_pdu.c:
```diff
@@ -726,7 +729,7 @@ void llcp_pdu_encode_cte_req(struct proc_ctx *ctx, struct pdu_data *pdu)
 
 void llcp_pdu_decode_cte_rsp(struct proc_ctx *ctx, const struct pdu_data *pdu)
 {
-	if (pdu->cp == 0U || pdu->cte_info.time == 0U) {
```
Not sure about the KConfigs - but will octet3.cte_info not be undef'ed for OPENISA
(sorry if I'm missing the obvious!)

cvinayak replied:
CTE is not supported in OPENISA, and these lines are conditionally compiled out.

erbr-ot replied:
That is a very satisfying answer ;-)

erbr-ot reviewed (approved):
LGTM
